### PR TITLE
ENH: Support jac=True option for constraints in `scipy.optimize.minimize`

### PR DESCRIPTION
**Pull request by jhelgert**
#### What does this implement/fix?

This PR adds support for a `jac=True` option for `scipy.optimize.minimize` constraints (NonlinearConstraint objects and dictionary constraints), see #13842.

### CODE REVIEW

**Review by tupui — COMMENTED** [review]
I am personally -1 for this. IMO it adds unnecessary possibilities/complexity.

**Comment by jhelgert** [comment]
> I am personally -1 for this. IMO it adds unnecessary possibilities/complexity.

I agree that it increases the complexity of properly handling the constraints. However, I can't really understand why you think it's unnecessary. The same option is already there for the objective gradient and allows to avoid duplicated computations of terms arising in both the objective and the gradient and thus reduces the solver runtimes for large problems. IMO, it's reasonable to allow the same option for the constraints for the same reason. I'm curious what the others think about it.

**Comment by andyfaff** [comment]
I think for SLSQP that the function and jacobian evaluations are not required at the same time, see lines 428 to [434](https://github.com/scipy/scipy/blob/master/scipy/optimize/slsqp.py#L434), so there's no real benefit for SLSQP at least.

**Comment by tupui** [comment]
> I agree that it increases the complexity of properly handling the constraints. However, I can't really understand why you think it's unnecessary. The same option is already there for the objective gradient and allows to avoid duplicated computations of terms arising in both the objective and the gradient and thus reduces the solver runtimes for large problems. IMO, it's reasonable to allow the same option for the constraints for the same reason. I'm curious what the others think about it.

I am just questioning such designs. With a class based approach, some LRU caching or other things, you could avoid any extra computations.

**Comment by andyfaff** [comment]
Also, when the `NonLinearConstraint`, etc, are made into a `PreparedConstraint` the constraint functions use [`VectorFunction`](https://github.com/scipy/scipy/blob/master/scipy/optimize/_constraints.py#L233), etc. Those `*VectorFunction` use a memoisation logic - no extra calculations for the same `x`.

**Comment by jhelgert** [comment]
Thanks for the fast responses! 

> I am just questioning such designs. With a class based approach, some LRU caching or other things, you could avoid any extra computations.
> Having a param which can be a boolean, string and callable... Polymorphism is nice, but starts to be a jungle IMHO.

I oriented myself to the status quo of the objective `jac` that accepts a string, boolean and a callable.

> Also, when the `NonLinearConstraint`, etc, are made into a `PreparedConstraint` the constraint functions use [`VectorFunction`](https://github.com/scipy/scipy/blob/master/scipy/optimize/_constraints.py#L233), etc. Those `*VectorFunction` use a memoisation logic - no extra calculations for the same `x`.

That's not what I mean by avoiding duplicated computations. `VectorFunction` avoids duplicated evaluations for the same `x` but there's still no way to cache intermediate results that arise both in the constraint function and the jacobian. Here's an example of what I mean:

``` python
import numpy as np
from scipy.optimize._differentiable_functions import VectorFunction
from scipy.optimize.optimize import MemoizeJac
from scipy.optimize._hessian_update_strategy import BFGS

# Generate random data
M = 1000
N = 1000
A = np.random.randn(M, N)
b = np.random.randn(M)


def con(x):
    return x.T @ A.T @ A @ x - 2*b.T @ A @ x + b.T @ b

def jac(x):
    return 2 * A.T @ A @ x - 2 * A.T @ b

def con_and_jac(x):
    # cache intermediate results
    AtA = A.T @ A
    AtAx = AtA @ x
    btA = b.T @ A
    Atb = btA.T
    # evaluate the constraint function and the jacobian
    con = x.T @ AtAx - 2 * btA @ x + b.T @ b
    jac = 2*AtAx - 2*Atb
    return con, jac

def test_evaluation(con_fun, con_jac):
    # generate random point
    x = np.random.randn(N)
    # Evaluate the constraint and the jacobian
    con_fun(x)
    con_jac(x)

x0 = np.random.randn(N)
vf = VectorFunction(con, x0, jac, hess=BFGS(), finite_diff_rel_step=None,
                    finite_diff_jac_sparsity=None, finite_diff_bounds=None, 
                    sparse_jacobian=None)
con_fun1 = vf.fun
con_jac1 = vf.jac

con_fun2 = MemoizeJac(con_and_jac)
con_jac2 = con_fun2.derivative
```

Here,  `A.T @ A` and `A.T @ A @ x` both arise in the function and the jacobian. By using `con_and_jac()` and allowing a `jac=True` option for the constraints, we only need to compute these expressions once for each evaluation:

``` python
In [33]: %timeit test_evaluation(con_fun1, con_jac1)
12.4 ms ± 287 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [34]: %timeit test_evaluation(con_fun2, con_jac2)
6.39 ms ± 49 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

> I think for SLSQP that the function and jacobian evaluations are not required at the same time, see lines 428 to [434](https://github.com/scipy/scipy/blob/master/scipy/optimize/slsqp.py#L434), so there's no real benefit for SLSQP at least.

Indeed, my bad! If interested, I can do some benchmarks for `trust-constr` and check whether there's a real benefit for large problems.

**Comment by tupui** [comment]
> Here, `A.T @ A` and `A.T @ A @ x` both arise in the function and the jacobian. By using `con_and_jac()` and allowing a `jac=True` option for the constraints, we only need to compute these expressions once for each evaluation:

Then just define this operation in a function and cache it with `functools.lru_cache`.

**Comment by andyfaff** [comment]
It's not really that possible to use `lru_cache` directly with numpy arrays. However, it would be possible to use a memoization approach similar to that done in `ScalarFunction` to store the intermediate results

**Comment by jhelgert** [comment]
> > Here, `A.T @ A` and `A.T @ A @ x` both arise in the function and the jacobian. By using `con_and_jac()` and allowing a `jac=True` option for the constraints, we only need to compute these expressions once for each evaluation:
> 
> Then just define this operation in a function and cache it with `functools.lru_cache`

This assumes that the exact jacobian and all common expressions are known. The main motivation of this PR is that it makes it very easy to use algorithmic differentiation utilities like [`jax.value_and_grad`](https://jax.readthedocs.io/en/latest/jax.html#jax.value_and_grad) to efficiently compute constraints and jacobians.

**Comment by tupui** [comment]
Still not a fan. I will let @andyfaff decide.

IMHO, I would also not blindly follow arguments like, it's easier with such and such tools (especially with things coming from Google which tend to come/die and do breaking changes all the time...). It does not necessarily mean that it's the right solution and we should think carefully before adding yet another different behaviour for a param.

But if we are certain that most/all tools providing a jacobian work like that, I would be +-0.

File: scipy/optimize/_differentiable_functions.py [hunk 1]
```diff
@@ -345,6 +345,12 @@ def __init__(self, fun, x0, jac, hess,
                              "be estimated using one of the quasi-Newton "
                              "strategies.")
 
+        if jac is True:
+            from scipy.optimize.optimize import MemoizeJac # no circular import
```
Comment (mdhaber) [inline scipy/optimize/_differentiable_functions.py:349]:
```suggestion
            from scipy.optimize.optimize import MemoizeJac  # no circular import
```

**Review by mdhaber — COMMENTED** [review]
I think I've run into this sort of need before, too. I'm usually pretty cautious about adding convenience functionality that the user could take care of, but I think this is complicated enough to do in userland that we should support it natively. There are other possibilities for _how_ to support it, but since there is precedent for the boolean idea, I think it's reasonable to use that approach again. 

@jhelgert The implementation is quite simple, so I appreciate that. One question, though: why the work for old-style constraints is done in `slsqp.py` rather than in `minimize`? (Is it because `slsqp` is currently the only method that would have old-style constraints and use the jacobian? I guess Cobyla doesn't?)

**Comment by jhelgert** [comment]
> One question, though: why the work for old-style constraints is done in slsqp.py rather than in minimize?
Is it because `slsqp` is currently the only method that would have old-style constraints and use the jacobian? I guess Cobyla doesn't?

Yeah, exactly. Cobyla is a derivative-free algorithm and doesn't use passed gradients and jacobians since the algorithm assumes that both are unknown. However, putting the work for the dict constraints directly into `minimize` seems reasonable to me since further algorithms (for constrained problems) added in the future would also benefit from it then.

**Comment by mdhaber** [comment]
> However, putting the work for the dict constraints directly into minimize seems reasonable to me since further algorithms (for constrained problems) added in the future would also benefit from it then.

Yup, that's what I was after. Please do that and I will take a closer look and probably merge.

File: scipy/optimize/tests/test_slsqp.py [hunk 2]
```diff
@@ -135,6 +135,14 @@ def test_minimize_unbounded_combined(self):
                        jac=True, method='SLSQP', options=self.opts)
         assert_(res['success'], res['message'])
         assert_allclose(res.x, [2, 1])
+    
+    def test_minimize_unbounded_con_combined(self):
+        # Minimize, method='SLSQP': unbounded, combined constraint and Jacobian
+        # constraint: 2*x*y + 2*x - x**2 - 2*y**2 >= 0
+        con = {'type': 'ineq', 'fun': self.fun_and_jac, 'jac': True}
+        res = minimize(self.fun_and_jac, [-1.0, 1.0], args=(-1.0, ),
+                       jac=True, constraints=(con,), method='SLSQP',
```
Comment (andyfaff) [inline scipy/optimize/tests/test_slsqp.py:144]: Can you create an appropriate test with, and without, the `func_and_grad` functionality implemented in this PR, to show that the number of function evaluations decreases somewhere?

**Review by andyfaff — CHANGES_REQUESTED** [review]

File: scipy/optimize/tests/test_slsqp.py (hunk 2 above)
Comment (jhelgert) [reply scipy/optimize/tests/test_slsqp.py:144]: I'm afraid that the `constr_nfev` and `constr_jfev` fields returned by 'trust-constr' that count the number of function/jacobian evaluations for each constraint won't change when using `jac=True` due to the logic of `MemoizeJac` (*). So I'm not really sure whether we should ignore this and if so, how could we test it properly?

To be precise, the constraint's jacobian function will still be called (and thus increases the `constr_jfev` counter) - the (only) difference is that it will use the jacobian computed in the previous evaluation of the constraint function. This can save a lot of computing time if the constraint function and the constraint jacobian share some expensive calculations.

The best I can think of would be something like this:

```python
class ConstraintFunctor:
    def __init__(self, fun, jac, con_and_jac=True):
        self.count_f = 0
        self.count_j = 0
        self.fun = fun
        self.jac = jac
        self.con_and_jac = con_and_jac
        
    def fun(self, x):
        self.count_f += 1
        if self.con_and_jac:
            return self.fun(x), self.jac(x)
        else:
            return self.fun(x)

    def jac(self, x):
        self.count_j += 1
        return self.jac(x)

cf1 = ConstraintFunctor(lambda x: x[0]**2, lambda x: 2*x[0], con_and_jac=True)
con1 = {'type': 'ineq', 'fun': cf1.fun, 'jac': cf1.jac}
res1 = minimize(x0=[-1.0, -1.0], constraints=(con1,), method="SLSQP")

cf2 = ConstraintFunctor(lambda x: x[0]**2, lambda x: 2*x[0], con_and_jac=False)
con2 = {'type': 'ineq', 'fun': cf2.fun, 'jac': True}
res2 = minimize(x0=[-1.0, -1.0], constraints=(con2,), method="SLSQP")
```

and then testing whether `cf1.count_f + cf1.count_j < cf2.count_f + cf2.count_j`. But this doesn't feel reasonable to me.

PS: (*) Afaik, the SLSQP algorithm doesn't count the function evaluations.
